docs/design/app/states.md

```diff
@@ -0,0 +1,301 @@
+objective to eliminate the need for a service module to interact directly with the merkle tree. This was done for four
+reasons. First, access to the full merkle tree allows a service to "dig around" in the tree and learn about other state
+and possibly depend on other state, outside its own *namespace*. For example, it would be awful for service "A" to find
+and depend upon state of service "B" by going through the merkle tree itself. The proper way to do this is to use the
+public API available on service "B". The second reason we try to abstract the details of the underlying merkle
+implementation is to give us a greater degree of flexibility for re-entrant cases like when the
+`hedera-smart-contract-service` uses a system contract to make a call to another service like `hedera-token-service`.
+Third, the merkle tree APIs are complex, and we wanted to localize that complexity in one package, and let the many
+service modules deal with simple POJO concepts and APIs, to decrease the likelihood of bugs and increase the
+testability of service modules.
+
+The final and strongest reason for introducing the classes in the `state` package is that it solves several of the
+more challenging problems in the services code base: deciding which copy of the merkle tree state to use, buffering
+all modifications to the merkle tree, and keeping track of all state read from the merkle tree. Let's go through each
+of these different core use cases and describe how the classes in `state` help with them.
+
+## Merkle Trees
+
+All persistent state is stored in a "merkle tree". A merkle tree is simply a tree structure used to reduce the amount
+of work needed to produce a "root hash" of all state in the tree (the tree caches the hashes of each node in the tree,
+so when a single leaf changes, we only have to compute the new hash of the leaf and the hash of each parent node in
+the tree to compute the new root hash). Performance can be improved by recomputing the root hash for multiple
+modifications: that is, re-hashing the tree after every modification is slower than re-hashing every N modifications,
+where N is a reasonably large value. We have found that when N is 10,000, it is faster-per-modification than when N is
+1.
+
+We also need to periodically compare "root hashes" of the state with all nodes in the network, to make sure that
+all nodes have matching state. Divergent state is a **very bad thing**, because it could mean that in the database
+of one machine you have an hbar balance of X, while on another machine you have an hbar balance of Y! Differences in
+the "state signature" of different machines is known as an ISS (Invalid State Signature).
+
+The hashgraph algorithm works on `Event`s, sorting them and bundling them together into `Round`s. Each `Event` has
+zero or more `PlatformTransaction`s. As each `Round` is produced, it is sent to the Hedera application to be handled.
```

Comment:
Ah, probably I could get away with that. At one point the documentation had to clarify which kind of transaction we mean, because each PlatformTransaction can have metadata, and we need to put the metadata onto the platform transaction, not onto the HAPI transaction. Having two things called transaction is really confusing, so I thought it better to try to be explicit so people get the right mental model